### PR TITLE
[Mining] Add the ability to create proof of work blocks on a proof of stake network.

### LINKED ISSUE #1614: Incorrect proof of work
Setting up a local network of two nodes, I had two nodes both mining, the first node was online first and mined 2 blocks, then I turned on the second node and it mined 1 block before it synced up with the other node. Then after a short while, the second node started complaining about "incorrect proof of work" (the node finds a new block).

```
Consensus error exception occurred in miner loop: Stratis.Bitcoin.Consensus.ConsensusErrorException: incorrect proof of work

at Stratis.Bitcoin.Consensus.ConsensusError.Throw() in C:\...\src\Stratis.Bitcoin\Consensus\ConsensusErrorException.cs:line 44

at Stratis.Bitcoin.Features.Consensus.Rules.CommonRules.BlockHeaderPowContextualRule.RunAsync(RuleContext context) in C:\...\src\Stratis.Bitcoin.Features.Consensus\Rules\CommonRules\BlockHeaderPowContextualRule.cs:line 33
```

The code at the line:

```
// Check proof of work.
if (header.Bits != context.NextWorkRequired)
```

Here is a zip of the data folder. This is built using the "main" Stratis network, but I have modified the node to not use checkpoints, no DNS seed and only my LAN nodes. The only modification to stratis.conf, is mine=1 and the address.

[StratisMain.zip](https://github.com/stratisproject/StratisBitcoinFullNode/files/2164378/StratisMain.zip)

I have attempted a few times to use this data folder when running StratisD from Visual Studio (Debugging), but then the problem does not occur. Only when I run from cmd, with "dotnet stratis.stratisd.dll".

I did a run with debug level on trace, and here is the log output from a quick startup and close. Search for "Validating new block" to find the interesting parts.

[node.zip](https://github.com/stratisproject/StratisBitcoinFullNode/files/2164421/node.zip)

### PR DESCRIPTION
**Pull request by fassadlr**
This PR fixes #1614 and #1625

### CODE REVIEW

**Comment by fassadlr** [comment]
I have successfully sync all the chainz.

**Comment by fassadlr** [comment]
Staking works as per normal 👍

File: src/Stratis.Bitcoin.Features.Miner/BlockProvider.cs [hunk 1]
```diff
@@ -30,6 +38,9 @@ public BlockTemplate BuildPosBlock(ChainedHeader chainTip, Script script)
         /// <inheritdoc/>
         public BlockTemplate BuildPowBlock(ChainedHeader chainTip, Script script)
         {
+            if (this.network.Consensus.IsProofOfStake)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Miner/BlockProvider.cs]: Ah you made the change here, I was planing to do it in the PowMiner.cs but I guess here is just as fine.

File: src/Stratis.Bitcoin.Features.Miner/PowMining.cs [hunk 2]
```diff
@@ -168,110 +168,181 @@ public void StopMining()
...
-                    // where it is possible to generate multiple blocks within one second.
-                    if (blockTemplate.Block.Header.Time <= chainTip.Header.Time)
-                        continue;
-                }
+                if (!BuildBlock(context))
+                    continue;
 
-                nExtraNonce = this.IncrementExtraNonce(blockTemplate.Block, chainTip, nExtraNonce);
-                Block block = blockTemplate.Block;
+                if (!MineBlock(context))
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Miner/PowMining.cs]: Reminder to add `this.` on all 3 mtheods

File: src/Stratis.Bitcoin.Features.Miner/PowMining.cs [hunk 3]
```diff
@@ -168,110 +168,181 @@ public void StopMining()
...
+
+            var chainedHeader = new ChainedHeader(context.BlockTemplate.Block.Header, context.BlockTemplate.Block.GetHash(), context.ChainTip);
+            if (chainedHeader.ChainWork <= context.ChainTip.ChainWork)
+                return false;
+
+            return true;
+        }
 
-                this.logger.LogInformation("Mined new {0} block: '{1}'.", BlockStake.IsProofOfStake(blockValidationContext.Block) ? "POS" : "POW", blockValidationContext.ChainedHeader);
+        private void ValidateMinedBlockWithConsensus(MineBlockContext context)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Miner/PowMining.cs]: I would suggest to merge this two methods `ValidateMinedBlockWithConsensus` `CheckValidationContext`
And call it `ValidateAndconnectBlock()`

File: src/Stratis.Bitcoin.Features.Miner/PowMining.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PowMining.cs]: 👍

File: src/Stratis.Bitcoin.Features.Miner/PowMining.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PowMining.cs]: 💯

**Pull request merged**